### ISSUE TITLE
async/import at runtime via HTTP/2

### ISSUE THREAD

**Issue by trusktr** [benjamn/reify#32]
I like the idea of async/import. In your description you talk about a bundler, which I assume means that code will be bundled on the server-side depending on static analysis.

But, ES6 modules in browsers will be able to load modules from URLs, so I think maybe the behavior needs to be described as something that happens at runtime. For example:
- The browser encounters an import statement in an executed function at runtime (in an async function, as you've described).
- The browser fetches that module (if it hasn't already when the async function was loaded) via HTTP/2
- When the module arrives over the network and is loaded, the browser keeps the HTTP/2 connection open, and with quick static analysis that happens while loading the module, the browser determines which set of modules to fetch next over the network through the same HTTP/2 connection, and fetches them.
- While loading the new set of modules, the browser will repeat that static-analysis process for each new module as it arrives and fetch the following sets of dependencies, repeating the process until the entire dependency graph for the original import statement is fetched and loaded.
- Finally, the browser can begin executing leaf-most modules (possibly while still fetching other dependencies that might be in deeper separate branches of the dep graph). Is JSPM doing something like this @guybedford?

It would become the responsibility of Webpack and similar libraries/bundlers to simply transpile and minify source files, but not actually bundle them, and to only send a minimal loader runtime to the client (or use the browser's loader when that is native), where the loader runtime expects the transpiled/minified/source-mapped files to exist at URLs that match the import statement module identifiers.

Just thought I'd throw this idea out there (and note that JSPM has something along these lines) which I think is what we should think about (for both Node and the browser). Now that [HTTP/2 support is getting broader](http://caniuse.com/#feat=http2), it might be nice for stacks like Meteor to take advantage of it, with a polyfilled loader API for now, and import sugar on top.

I personally don't have experience making a bundler or loader, and have only used them, but to me bundling sounds like a pain point for performance and is based on concepts of a time before Node, when bundling was the best thing to do in a browser.

If/when ES6 modules are native in browsers (and in Node, but in that case modules may come not just from the network but from the filesystem, or possibly _only_ the filesystem), then piping minified source files through HTTP/2 just seems like the best way to do it conceptually, completely dynamically where the runtime of an app can cause completely different sets of module files to be fetched depending on what the user does (or while the app is idle as you mentioned Ben, in order to speed things up and so that the import statement finishes sooner when it actually gets executed).

If something like this were implemented in Meteor, it could be possible to incorporate a feature where the client-side loader can detect (over time, after multiple clients have accessed an app) which modules are more likely to be needed by any given user in order to load those when the app is idle before loading less-commonly-used modules. Over some time, the app's load time (for common routes or views for example) will see improvements. I would also argue for there to be an option to turn this optimizing behavior, as some people might like for module-loading to happen only when import statements are encountered, and perhaps they want to implement their own module loading optimizations by simply using async/import and handing those dependencies to their app as they design themselves.

As mentioned, I don't have actual experience in making loaders/bundlers, but I really like the sound of this idea and would love to be the end user of such a thing. 😆

**Comment by trusktr** [comment]
Problem: import statements are transpiled on the server-side right now (in Meteor), so files being loaded via HTTP/2 would not have import statements, making static analysis more difficult. Maybe such a polyfilled system (in the meantime, while we wait for ES6 modules to be native), would just not transpile import statements, transpiling everything else server-side, but leaving transpilation of import statements for the client-side before loading and evaluating the module?

Such a system might also be groundwork for what Browsers would end up actually doing (and Node) but without needing to transpile.

**Comment by benjamn** [comment]
These are great points, @trusktr. I 100% agree we have to think through exactly how module fetching is going to work in the browser, without the luxury of a server-side bundling tool, if we want to have any confidence that nested `import` declarations are a coherent/implementable idea.

Perhaps the highest-leverage way forward is to implement support for nested `import` declarations in JSPM, since they're tackling all the challenges you're describing, except they're not (yet) doing nested `import`s. The `System.register` API isn't ideal for nested `import`s (as I discuss [here](https://github.com/benjamn/reify/blob/master/WHY_NEST_IMPORTS.md#nested-import-declarations-are-difficult-to-transpile)), but I think we can make it work.

**Comment by benjamn** [comment]
Before someone from JSPM claims they support nested `import`s via `System.import`, what I'm talking about are *synchronous* nestable `import` declarations, which do not desugar to `await System.import(...)`, though of course `System.import` still works the same as always.

Specifically, you'd use `System.import` to import your entry point module (the way JSPM recommends), and `System.import` would use HTTP/2 to fetch the whole transitive dependency tree (without evaluating any modules, yet), and only after fetching those dependencies would it begin evaluating entry point module. Now, since the source of all modules imported anywhere in the dependency tree is immediately available, `import` declarations can be implemented synchronously. Fetching all dependencies up-front sounds scary, but it's not that different from what would happen if you manually hoisted all your `import` statements to the top level.

**Comment by benjamn** [comment]
If you have a nested call to `System.import("./module")` somewhere in your code, the simplest implementation would be to treat it exactly like a top-level `System.import` call. However, when the argument is a string literal, it's possible that the original `System.import` call could decide to fetch `./module` preemptively, instead of waiting for `System.import("./module")` to be called.

**Comment by jdalton** [comment]
Is this more of a Meteor issue? A browser runnable module bit seems a bit out of scope for reify.

**Comment by benjamn** [comment]
My current perspective: Reify works great with the https://www.npmjs.com/package/install package, which leaves the implementation of module fetching up to the [`install.fetch`](https://github.com/benjamn/install/blob/fa332017a584b3b643ee999e44f5e6146e1bdba0/install.js#L160-L165) callback. In other words, this library is at least two steps removed from the details of how modules should be fetched, and Meteor 1.5 happens to [fetch modules over a `WebSocket`](https://github.com/meteor/meteor/blob/ff53a9115256c3bd8ea78c904ef8a7a43a29a4da/packages/dynamic-import/client.js#L17). These issues are all quite a bit more concrete than when @trusktr first opened this issue, so, with that clarity, I'm confident we can move/continue this discussion elsewhere, and let Reify remain unconcerned with what happens at the network level.
